### ISSUE TITLE
The first topicref to a ditabase file always pulls in the first topic

### ISSUE THREAD

**Issue by mnordseth** [dita-ot/dita-ot#1904]
I have a ditabase file with several topics:

``` xml
<!DOCTYPE dita PUBLIC "-//OASIS//DTD DITA Composite//EN" "ditabase.dtd" >
<dita>
  <reference id="t1">
    <title>Topic 1</title>
  </reference>

  <reference id="t2">
    <title>Topic 2</title>
  </reference>

  <reference id="t3">
    <title>Topic 3</title>
  </reference>
</dita>
```

And a map that pulls in those topics:

``` xml
<?xml version='1.0' encoding='UTF-8'?>
<!DOCTYPE map PUBLIC "-//OASIS//DTD DITA Map//EN" "map.dtd">
<map>
  <topicref href="test.dita#t2"/>
  <topicref href="test.dita#t1"/>
  <topicref href="test.dita#t3"/>
</map>
```

The output will always put topic t1 first, while the remaining topics are in the expected order.

**Comment by jelovirt** [comment]
Which DITA-OT version, which transtype?

**Comment by mnordseth** [comment]
DITA-OT 1.8.5, transtype pdf, pdf2 and pdf5 (AntennaHouse).

On Thu, Apr 30, 2015 at 11:03 AM, Jarno Elovirta <notifications@github.com>
wrote:

> Which DITA-OT version, which transtype?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dita-ot/dita-ot/issues/1904#issuecomment-97712784>.
>



-- 
Magnus Nordseth

**Comment by robander** [comment]
PDF has always had problems with the composite topic. Per the spec, a reference to the file (like `href="all.dita"` should bring in all of the topics, but has only ever pulled in the first - think we have a very old issue open for that. I'm pretty sure this comes from a problem in how the merged file is created.

**Comment by eerohele** [comment]
[Fixtures](https://github.com/eerohele/dita-ot-issues/tree/master/fixtures/1904).

I believe the issue is that when determining the ID of the first topic in the topicref target file, [`MergeMapParser.java`](https://github.com/dita-ot/dita-ot/blob/c682aef06f7d58c5ad2a159a0c2a98298dff4b4c/src/main/java/org/dita/dost/reader/MergeMapParser.java#L210) always gets the ID of the first topic in the target file, rather than using the ID defined in the `@href` attribute of the topicref itself.

So, to use my files as an example, `MergeMapParser.java` reads `root.ditamap` and when it comes across `<topicref href="topic1.dita#topic2"/>`, it opens `topic1.dita`, looks inside and sees that `#topic1` is the first topic in that file.

It then decides `#topic1` is the ID of the first topic in that file, rather than using `#topic2`, which is defined in the `@href` attribute of the `<topicref>` element itself.

One possible fix is something like this:

```diff
diff --git a/src/main/java/org/dita/dost/reader/MergeMapParser.java b/src/main/java/org/dita/dost/reader/MergeMapParser.java
index cf5df80..f09af06 100644
--- a/src/main/java/org/dita/dost/reader/MergeMapParser.java
+++ b/src/main/java/org/dita/dost/reader/MergeMapParser.java
@@ -56,6 +56,8 @@ public final class MergeMapParser extends XMLFilterImpl {
     private File dirPath = null;
     private File tempdir = null;
 
+    private boolean isFirstTopicRef = true;
+
     private final Stack<String> processStack;
     private int processLevel;
     private final ByteArrayOutputStream topicBuffer;
@@ -207,7 +209,20 @@ public final class MergeMapParser extends XMLFilterImpl {
                         final File f = new File(dirPath, toFile(p).getPath());
                         if (f.exists()) {
                             topicParser.parse(toFile(p).getPath(), dirPath);
-                            final String fileId = topicParser.getFirstTopicId();
+                            final String fileId;
+
+                            // If this is the first topicref and its @href
+                            // attribute has an ID, use that ID as the first
+                            // topic ID.
+                            //
+                            // Otherwise, use the ID of the first topic in
+                            // the topicref target.
+                            if (isFirstTopicRef && attValue.getFragment() != null) {
+                                fileId = util.getIdValue(attValue);
+                            } else {
+                                fileId = topicParser.getFirstTopicId();
+                            }
+
                             util.addId(attValue, fileId);
                             if (attValue.getFragment() != null) {
                                 util.addId(stripFragment(attValue), fileId);
@@ -227,6 +242,8 @@ public final class MergeMapParser extends XMLFilterImpl {
                     }
                 XMLUtils.addOrSetAttribute(atts, ATTRIBUTE_NAME_HREF, attValue.toString());
             }
+
+            isFirstTopicRef = false;
         }
         getContentHandler().startElement(uri, localName, qName, atts != null ? atts : attributes);
     }
```

@jelovirt: Let me know if you'd like a PR.

**Comment by jelovirt** [comment]
This will not work when the map contains multiple topicrefs so ditabase files. The implementation cannot rely on first occurrence for all topicrefs.

**Comment by robander** [comment]
I ran into this behavior again when trying to work on #1077 last week (currently our oldest open issue). Both issues are related, though they address different bugs.

**Comment by markgif** [comment]
I think I ran into this with a different setup than the one @mnordseth shows. Here is my ditabase file with two embedded topics:

    <dita>
      <topic id="t1">
        <title>Topic title BAD</title>
        <body>
          <p>Here is topic text.</p>
        </body>
      </topic>
      <concept id="c1">
        <title>Concept 1 title</title>
        <conbody>
          <p>Concept text.</p>
        </conbody>
      </concept>
    </dita>

Here is my map, which references the ditabase file itself, not a topic inside the ditabase:

    <map>
     <title>map</title>
      <topicref href="ditabase-test.dita"/>
    </map>

With the OT 1.8.5 pdf2 transtype, only the first topic appears in the PDF. But if I embed the second topic inside the first one, then it works in the PDF:

    <dita>
      <topic id="t1">
        <title>Topic title BAD</title>
        <body>
          <p>Here is topic text.</p>
        </body>
        <concept id="c1">
          <title>Concept 1 title</title>
          <conbody>
            <p>Concept text.</p>
          </conbody>
        </concept>
      </topic>
    </dita>

**Comment by robander** [comment]
The issue here was the one Eero described (grabbing the first root topic's ID for all cases), and was resolved by #2679. Tagging this with 2.5 and closing.
